plugins/randomforest_procedure.cc

```diff
@@ -205,79 +233,118 @@ run(const ProcedureRunConfig & run,
-            std::random_shuffle(tr_ex_nums.begin(), tr_ex_nums.end(), myrng);
-            for (unsigned i = 0;  i < numRows * trainprop;  ++i)
-                in_training[tr_ex_nums[i]] = 1.0;
+            size_t numPartitions = std::min<size_t>(numRows, 32);
 
-            distribution<float> example_weights(numRows);
+            distribution<float> trainingWeights(numRows);
+            std::atomic<size_t> numNonZero(0);
+            
+            // Parallelize the setup, since the slow part is random number
+            // generation and we set up each bag independently
+            auto doPartition = [&] (size_t i)
+            {
+                RandomForestRNG myrng(38341 + i);
+
+                size_t first = (numRows / numPartitions) * i;
+                size_t last = (numRows / numPartitions) * (i + 1);
+                if (last > numRows || i == numPartitions)
+                    last = numRows;
+
+                boost::mt19937 rng(245 + (bag * numPartitions) + i);
+
+                size_t numRowsInPartition = last - first;
+
+                distribution<float> in_training(numRowsInPartition);
+                vector<int> tr_ex_nums(numRowsInPartition);
+                std::iota(tr_ex_nums.begin(), tr_ex_nums.end(), 0);
+                std::random_shuffle(tr_ex_nums.begin(), tr_ex_nums.end(), myrng);
+                
+                for (unsigned i = 0;  i < numRowsInPartition * trainprop;  ++i)
+                    in_training[tr_ex_nums[i]] = 1.0;
+
+                distribution<float> example_weights(numRowsInPartition);
+
+                // Generate our example weights.
+                for (unsigned i = 0;  i < numRowsInPartition;  ++i)
+                    example_weights[myrng(numRowsInPartition)] += 1.0;
+
+                size_t partitionNumNonZero = 0;
+                double totalTrainingWeights = 0;
+                for (unsigned i = 0;  i < numRowsInPartition;  ++i) {
+                    float wt = in_training[i] * example_weights[i];
+                    trainingWeights[first + i] = wt;
+                    partitionNumNonZero += (wt != 0);
+                    totalTrainingWeights += wt;
+                }
 
-            // Generate our example weights.
-            for (unsigned i = 0;  i < numRows;  ++i)
-                example_weights[myrng(numRows)] += 1.0;
+                SIMD::vec_scale(trainingWeights.data() + first,
+                                1.0 / (totalTrainingWeights * numPartitions),
+                                trainingWeights.data() + first,
+                                numRowsInPartition);
+                numNonZero += partitionNumNonZero;
+            };
 
-            distribution<float> training_weights
-                = in_training * example_weights;
+            parallelMap(0, numPartitions, doPartition);
 
-            training_weights.normalize();
+            cerr << "bag " << bag << " weight generation took "
```

Comment:
I will leave this one out for the moment; the rest of the procedure still logs to cerr.  https://datacratic.atlassian.net/browse/MLDB-2014